tools/lint/common.c

```diff
@@ -0,0 +1,586 @@
+            ptr[-1] = ' ';
+
+            if ((ptr[len - 1] == qmark) && (ptr[len - 2] != '\\')) {
+                /* end of quotation */
+                qmark = 0;
+                /* shorten the argument by the terminating quotation mark */
+                ptr[len - 1] = '\0';
+            }
+            continue;
+        }
+
+        /* another token in cmdline */
+        ++count;
+        r = realloc(vector, (count + 1) * sizeof *vector);
+        if (!r) {
+            YLMSG_E("Memory allocation failed (%s:%d, %s),", __FILE__, __LINE__, strerror(errno));
+            free(vector);
+            return -1;
+        }
+        vector = r;
+        vector[count - 1] = ptr;
+
+        if ((ptr[0] == '"') || (ptr[0] == '\'')) {
+            /* remember the quotation mark to identify end of quotation */
+            qmark = ptr[0];
+
+            /* move the remembered argument after the quotation mark */
+            ++vector[count - 1];
+
+            /* check if the quotation is terminated within this token */
+            if ((ptr[len - 1] == qmark) && (ptr[len - 2] != '\\')) {
+                /* end of quotation */
+                qmark = 0;
+                /* shorten the argument by the terminating quotation mark */
+                ptr[len - 1] = '\0';
+            }
+        }
+    }
+    vector[count] = NULL;
+
+    *argc_p = count;
+    *argv_p = vector;
+
+    return 0;
+}
+
+int
+get_format(const char *filename, LYS_INFORMAT *schema, LYD_FORMAT *data)
+{
+    char *ptr;
+    LYS_INFORMAT informat_s;
+    LYD_FORMAT informat_d;
+
+    /* get the file format */
+    if ((ptr = strrchr(filename, '.')) != NULL) {
+        ++ptr;
+        if (!strcmp(ptr, "yang")) {
+            informat_s = LYS_IN_YANG;
+            informat_d = 0;
+        } else if (!strcmp(ptr, "yin")) {
+            informat_s = LYS_IN_YIN;
+            informat_d = 0;
+        } else if (!strcmp(ptr, "xml")) {
+            informat_s = 0;
+            informat_d = LYD_XML;
+        } else if (!strcmp(ptr, "json")) {
+            informat_s = 0;
+            informat_d = LYD_JSON;
+        } else {
+            YLMSG_E("Input file \"%s\" in an unknown format \"%s\".\n", filename, ptr);
+            return 0;
+        }
+    } else {
+        YLMSG_E("Input file \"%s\" without file extension - unknown format.\n", filename);
+        return 1;
+    }
+
+    if (informat_d) {
+        if (!data) {
+            YLMSG_E("Input file \"%s\" not expected to contain data instances (unexpected format).\n", filename);
+            return 2;
+        }
+        (*data) = informat_d;
+    } else if (informat_s) {
+        if (!schema) {
+            YLMSG_E("Input file \"%s\" not expected to contain schema definition (unexpected format).\n", filename);
+            return 3;
+        }
+        (*schema) = informat_s;
+    }
+
+    return 0;
+}
+
+int
+print_list(struct ly_out *out, struct ly_ctx *ctx, LYD_FORMAT outformat)
```

Comment:
Functions such as this one are called only once in the specific command so why not make the function static and put it there? I would find that better organized.

Reply:
It's called twice - typically these functions are called from `cmd_*` file and from `main.ni.c` - libyang has 2 modes, interactive and non-interactive. I hope that whenever possible, I have unified the code to avoid duplication. What must be solved individually is command line processing, but executing command is in many cases common.

Reply:
Ah, okay, did not realize that. Then it makes sense.